### ISSUE TITLE
Java proxies sometimes not bound quickly enough

### ISSUE THREAD

**Issue by cheald** [jruby/jruby#3166]
It appears that in some cases, Java class proxies can be made available before they're fully bound. We are using the Stanford NLP libraries to perform entity extraction on text, with the process:

1. Create an `Annotator`
2. Pass text to the `Annotator`, which produces an `Annotation` instance
3. Call `.get(TokensAnnotation)` to get a `List<CoreLabel>`
4. Iterate the `List<CoreLabel>` and for each `CoreLabel` call `#original_text`
```ruby
      class EntityExtractor < Component::Service
        java_import "edu.stanford.nlp.ling.CoreAnnotations$TokensAnnotation"

        include_package "edu.stanford.nlp.pipeline"
        java_import "java.util.Properties"
        java_import "intoxicant.analytics.coreNlp.StopwordAnnotator"

        # Binds an RMQ consumer that receives jobs
        process do |msg|
          if msg["data"]["type"] == "html"
            page = Page.id_or_url(msg["id"], msg["url"])
            process msg, page if page
          end
        end

        @semaphore = Mutex.new

        def self.pipeline
          @semaphore.synchronize do
            @pipeline ||= begin
              props = Properties.new.tap do |p|
                p.set_property "annotators", "tokenize, ssplit, stopword, pos, lemma, ner"
                p.set_property "ner.applyNumericClassifiers", "false"
                p.set_property "ner.useSUTime", "false"
                p.set_property "parse.maxlen", "80" # See https://mailman.stanford.edu/pipermail/java-nlp-user/2014-November/006483.html
                p.set_property "customAnnotatorClass.stopword", "intoxicant.analytics.coreNlp.StopwordAnnotator"
              end
              StanfordCoreNLP.new props
            end
          end
        end

        def process(msg, page)
          entities = extract_entities annotate(msg["data"]["article"])
        end

        def annotate(article)
          Annotation.new(article).tap {|doc| self.class.pipeline.annotate(doc) }
        end

        def extract_entities(document)
          document.get(TokensAnnotation).map {|e| [e.original_text, e.ner] } # < -- undefined method `original_text' for #<Java::EduStanfordNlpLing::CoreLabel:0x412ce5ad>
          # ...
        end
      end
```
However, once per app restart (across several dozen instances of the application) 1-2 instances produce an error like:

    undefined method `original_text' for #<Java::EduStanfordNlpLing::CoreLabel:0x412ce5ad>

Future invocations work just fine.

Per IRC discussion, this sounds like an error with the proxy class being made available before it's fully bound.

**Comment by kares** [comment]
these are tricky to fix but reproducing helps, which seems like we won't be able to in this case ;(
... I assume all instances in `List<CoreLabel>` are actually of class `CoreLabel` (no sub-classes) ?
... also for the record what JRuby version are you using ... latest of 1.7.x (1.7.21) if not pls try.

**Comment by cheald** [comment]
Yes, they are actually CoreLabel instances. This is deployed against jruby-1.7.19 - I'll try an upgrade to .21 tomorrow.

**Comment by kares** [comment]
@cheald that is actually good news ... this should be fixed than since **1.7.20** due #2014 and #1621 so pls do

**Comment by cheald** [comment]
I upgraded to 1.7.21 last night and haven't seen this recur yet. I'll reopen the issue if it does.
